src/crab/type_domain.hpp

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <unordered_map>
+
+#include "crab/abstract_domain.hpp"
+#include "crab/cfg.hpp"
+#include "linear_constraint.hpp"
+#include "string_constraints.hpp"
+
+namespace crab {
+
+enum class region {
+	T_CTX,
+	T_STACK,
+	T_PACKET,
+	T_SHARED
+};
+
+struct ptr_no_off_t {
+    region r;
+
+    ptr_no_off_t() = default;
+    ptr_no_off_t(const ptr_no_off_t &) = default;
+    ptr_no_off_t(ptr_no_off_t &&) = default;
+    ptr_no_off_t &operator=(const ptr_no_off_t &) = default;
+    ptr_no_off_t &operator=(ptr_no_off_t &&) = default;
+    ptr_no_off_t(region _r) : r(_r) {}
+
+    friend std::ostream& operator<<(std::ostream& o, const ptr_no_off_t& p);
+    friend bool operator==(const ptr_no_off_t& p1, const ptr_no_off_t& p2);
+    friend bool operator!=(const ptr_no_off_t& p1, const ptr_no_off_t& p2);
+};
+
+struct ptr_with_off_t {
+    region r;
+    int offset;
+
+    ptr_with_off_t() = default;
+    ptr_with_off_t(const ptr_with_off_t &) = default;
+    ptr_with_off_t(ptr_with_off_t &&) = default;
+    ptr_with_off_t &operator=(const ptr_with_off_t &) = default;
+    ptr_with_off_t &operator=(ptr_with_off_t &&) = default;
+    ptr_with_off_t(region _r, int _off) : r(_r), offset(_off) {}
+
+    friend std::ostream& operator<<(std::ostream& o, const ptr_with_off_t& p);
+    friend bool operator==(const ptr_with_off_t& p1, const ptr_with_off_t& p2);
+    friend bool operator!=(const ptr_with_off_t& p1, const ptr_with_off_t& p2);
+};
+
+using ptr_t = std::variant<ptr_no_off_t, ptr_with_off_t>;
+using register_t = uint8_t;
+using location_t = boost::optional<std::pair<label_t, uint32_t>>;
+
+struct reg_with_loc_t {
```

Comment:
I don't like that `r` and `loc` are public. You do it so that the friend methods can access to them. Don't do that. In this case you can define a method `void write(std::ostream &o) const` that prints the stuff and then `operator<<` just calls `write`.